plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/panel/GitPanelViewImpl.java

```diff
@@ -0,0 +1,139 @@
+  @UiField(provided = true)
+  final GitLocalizationConstant locale;
+
+  @UiField(provided = true)
+  Tree repositoriesList;
+
+  private final RepositoryNodeFactory repositoryNodeFactory;
+
+  @Inject
+  public GitPanelViewImpl(
+      GitPanelViewImplUiBinder uiBinder,
+      RepositoryNodeFactory repositoryNodeFactory,
+      GitResources gitResources,
+      GitLocalizationConstant locale) {
+    this.repositoryNodeFactory = repositoryNodeFactory;
+    this.gitResources = gitResources;
+    this.locale = locale;
+
+    createRepositoriesList();
+    setContentWidget(uiBinder.createAndBindUi(this));
+  }
+
+  private void createRepositoriesList() {
+    NodeStorage nodeStorage = new NodeStorage();
+    NodeLoader nodeLoader = new NodeLoader();
+    repositoriesList = new Tree(nodeStorage, nodeLoader);
+    repositoriesList.getElement().getStyle().setProperty("maxHeight", "200px");
+    repositoriesList
+        .getNodeStorage()
+        .addSortInfo(new StoreSortInfo(Comparator.comparing(Node::getName), SortDir.ASC));
+    SelectionModel selectionModel = repositoriesList.getSelectionModel();
+    selectionModel.setSelectionMode(SelectionModel.Mode.SINGLE);
+    selectionModel.addSelectionChangedHandler(
+        event -> {
+          Node node = getFirst(event.getSelection(), null);
+          if (node == null) {
+            delegate.onRepositorySelectionChanged(null);
+          } else {
+            delegate.onRepositorySelectionChanged(node.getName());
+          }
+        });
+  }
+
+  @Override
+  public void setChangesPanelView(ChangesPanelView changesPanelView) {
+    this.changesPanel.add(changesPanelView);
+  }
+
+  @Override
+  public void addRepository(String repository) {
+    repositoriesList.getNodeStorage().add(repositoryNodeFactory.newRepositoryNode(repository, 0));
+  }
+
+  @Override
+  public void removeRepository(String repository) {
+    repositoriesList.getNodeStorage().remove(findNode(repository));
+  }
+
+  @Override
+  public void renameRepository(String oldRepositoryName, String newRepositoryName) {
+    RepositoryNode node = findNode(oldRepositoryName);
+    if (node != null) {
+      NodeStorage nodeStorage = repositoriesList.getNodeStorage();
+      nodeStorage.remove(node);
+      nodeStorage.add(
+          repositoryNodeFactory.newRepositoryNode(newRepositoryName, node.getChanges()));
+    }
+  }
+
+  @Override
+  public void updateRepositoryChanges(String repository, int changes) {
+    RepositoryNode node = findNode(repository);
+    if (node != null) {
+      node.setChanges(changes);
+      repositoriesList.refresh(node);
+    }
+  }
+
+  @Override
+  public String getSelectedRepository() {
+    List<Node> selection = repositoriesList.getSelectionModel().getSelectedNodes();
+    if (selection.isEmpty()) {
+      return null;
+    }
+    return selection.get(0).getName();
+  }
+
+  private RepositoryNode findNode(String repositoryName) {
+    for (Node node : repositoriesList.getNodeStorage().getAll()) {
```

Comment:
It is possible to use some key to find a node?

Reply:
Yes, but the key isn't reflect bind node, just a hash. In such case we have to store all info in additional map. Note, that quantity of repositories is usually small.